internal/lcp/lcp.go

```diff
@@ -0,0 +1,237 @@
+		return nil, errors.New("not an LCP packet")
+	}
+	b = b[2:] // Advance past PPP frame, to start of LCP packet.
+
+	ret := &Packet{
+		Code: Code(b[0]),
+		ID:   b[1],
+	}
+	pktLen := int(binary.BigEndian.Uint16(b[2:4]))
+	// Note that pktLen == len(b) is not required. PPP explicitly
+	// allows trailing padding to be inserted by the framing layer,
+	// and requires that inner protocols like LCP should gracefully
+	// ignore any padding.
+	//
+	// So, we only check if the declared packet length is
+	// nonsensically short, or overflows the total packet size.
+	if pktLen < 4 {
+		return nil, io.ErrUnexpectedEOF
+	}
+	if pktLen > len(b) {
+		return nil, io.ErrUnexpectedEOF
+	}
+
+	b = b[4:pktLen] // Advance to packet payload
+
+	switch ret.Code {
+	case typeConfigureRequest, typeConfigureAck, typeConfigureNak, typeConfigureReject:
+		opts, err := parseLCPOptions(b)
+		if err != nil {
+			return nil, err
+		}
+		for opt, val := range opts {
+			switch opt {
+			case optionMRU:
+				if len(val) != 2 {
+					return nil, errUnexpectedLen
+				}
+				ret.MRU = binary.BigEndian.Uint16(val)
+				delete(opts, opt)
+			case optionAuthProto:
+				if len(val) < 2 {
+					return nil, io.ErrUnexpectedEOF
+				}
+				ret.AuthProto = binary.BigEndian.Uint16(val[:2])
+				if ret.AuthProto == 0xc223 {
+					if len(val) != 3 {
+						return nil, errUnexpectedLen
+					}
+					ret.CHAPAlgorithm = val[2]
+				} else if len(val) != 2 {
+					return nil, errUnexpectedLen
+				}
+				delete(opts, opt)
+			case optionMagic:
+				if len(val) != 4 {
+					return nil, errUnexpectedLen
+				}
+				ret.Magic = binary.BigEndian.Uint32(val)
+				delete(opts, opt)
+			}
+		}
+		ret.UnknownOptions = opts
+
+	case typeProtocolReject:
+		if len(b) < 2 {
+			return nil, io.ErrUnexpectedEOF
+		}
+		ret.RejectedProtocol = binary.BigEndian.Uint16(b[:2])
+		b = b[2:]
+		fallthrough
+	case typeTerminateRequest, typeTerminateAck, typeCodeReject:
+		ret.Data = b
+
+	case typeEchoRequest, typeEchoReply, typeDiscardRequest:
+		if len(b) < 4 {
+			return nil, errors.New("packet too short")
+		}
+		ret.Magic = binary.BigEndian.Uint32(b[:4])
+		ret.Data = b[4:]
+
+	default:
+		return nil, fmt.Errorf("unknown LCP packet type %x", ret.Code)
+	}
+
+	return ret, nil
+}
+
+func parseLCPOptions(b []byte) (map[uint8][]byte, error) {
+	ret := map[uint8][]byte{}
+
+	for len(b) > 0 {
+		if len(b) < 2 {
+			return nil, errors.New("trailing garbage at end of packet")
+		}
+		optionType, optionLen := b[0], int(b[1])
+		if optionLen < 2 {
+			return nil, fmt.Errorf("option length %d for option %d is too short", optionLen, optionType)
+		}
+		if optionLen > len(b) {
+			return nil, fmt.Errorf("option length %d for option %d overflows packet", optionLen, optionType)
+		}
+		ret[optionType] = b[2:optionLen]
+		b = b[optionLen:]
+	}
+
+	return ret, nil
+}
+
+// Bytes serializes an LCP packet into a PPP frame for transmission.
+func (p *Packet) Bytes() []byte {
+	var out bytes.Buffer
```

Comment:
You could make this less garbage-y by making this an `AppendSerialized(b []byte) []byte` method.

So callers could give you scratch buffer capacity to work with. You then just append as needed, then return the final payload. You can still do your fill-in-the-length trick.

But then callers doing a bunch of these can setup the buffer once and reuse it.